store/src/java/com/zimbra/cs/service/AuthProvider.java

```diff
@@ -820,13 +820,44 @@ public static AuthToken getAuthToken(Account account, Usage usage, TokenType tok
         throw AuthProviderException.FAILURE(String.format("cannot get authtoken from account ", account.getName()));
     }
 
+    public static AuthToken getAuthToken(Account account, boolean isAdmin, Usage usage, TokenType tokenType) throws AuthProviderException {
+        List<AuthProvider> providers = getProviders();
+        AuthProviderException authProviderExp = null;
+
+        for (AuthProvider ap : providers) {
+            try {
+                AuthToken at = ap.authToken(account, isAdmin, usage, tokenType);
+                if (at == null) {
+                    authProviderExp = AuthProviderException.FAILURE(String.format("auth provider %s returned null", ap.getName()));
+                } else {
+                    return at;
+                }
+            } catch (AuthProviderException e) {
+                if (e.canIgnore()) {
+                    logger().debug("auth provider failure %s : %s", ap.getName(), e.getMessage());
+                } else {
+                    authProviderExp = e;
```

Comment:
Continuing using for does not make sense if we get an exception here. There are many ways to improve at it, one way will be to throw an exception from here, or break it and throw it afterwards.